core.js

```diff
@@ -372,114 +412,324 @@ exports.init = function (sbot, config) {
+      // prettier-ignore
+      return cb(new Error('addTransaction() failed because feed format ' + feedFormat.name + ' does not support validateBatch'))
+    }
+    if (!feedFormat.validateOOOBatch) {
+      // prettier-ignore
+      return cb(new Error('addTransaction() failed because feed format ' + feedFormat.name + ' does not support validateOOOBatch'))
     }
 
     onceWhen(
       stateFeedsReady,
       (ready) => ready === true,
       () => {
-        if (content.recps) {
-          try {
-            content = encryptContent(keys, content)
-          } catch (ex) {
-            return cb(ex)
-          }
+        const done = multicb({ pluck: 1 })
+
+        if (nativeMsgs.length > 0) {
+          const feedId = feedFormat.getFeedId(nativeMsgs[0])
+          const prevNativeMsg = state.get(feedId)
+          feedFormat.validateBatch(hmacKey, nativeMsgs, prevNativeMsg, done())
+        } else {
+          done()(null, [])
         }
-        const latestKVT = state[keys.id]
-        const msgVal = validate.create(
-          latestKVT ? { queue: [latestKVT] } : null,
+
+        feedFormat.validateOOOBatch(hmacKey, oooNativeMsgs, done())
+
+        done((err) => {
+          // prettier-ignore
+          if (err) return cb(clarify(err, 'validation in addTransaction() failed'))
+
+          const msgIds = nativeMsgs.map((nMsg) => feedFormat.getMsgId(nMsg))
+          const oooMsgIds = oooNativeMsgs.map((nMsg) =>
+            feedFormat.getMsgId(nMsg)
+          )
+
+          if (nativeMsgs.length > 0) {
+            const lastIndex = nativeMsgs.length - 1
+            const nativeMsg = nativeMsgs[lastIndex]
+            const feedId = feedFormat.getFeedId(nativeMsg)
+            state.update(feedId, nativeMsg)
+          }
+
+          const allMsgIds = [].concat(msgIds, oooMsgIds)
+          const allMsgVals = []
+            .concat(nativeMsgs, oooNativeMsgs)
+            .map((nMsg) => feedFormat.fromNativeMsg(nMsg, opts.encoding))
+
+          log.addTransaction(
+            allMsgIds,
+            allMsgVals,
+            opts.encoding,
+            (err, kvts) => {
+              if (err)
+                return cb(clarify(err, 'addTransaction() failed in the log'))
+              if (kvts.length !== allMsgIds.length) {
+                // prettier-ignore
+                return cb(new Error('addTransaction() failed due to mismatched message count'))
+              }
+
+              for (let i = 0; i < kvts.length; ++i) {
+                post.set(kvts[i])
+                onMsgAdded.set({
+                  kvt: kvts[i],
+                  nativeMsg: allMsgVals[i],
+                  feedFormat: feedFormat.name,
+                })
+              }
+              cb(null, kvts)
+            }
+          )
+        })
+      }
+    )
+  }
+
+  function create(opts, cb) {
+    const guard = guardAgainstDuplicateLogs('create()')
+    if (guard) return cb(guard)
+
+    const keys = opts.keys || config.keys
```

Comment:
> Maybe require keys to be defined if format is not classic?

Yes, we can do that, but additionally we can add a check: `keys.id` has to pass `feedFormat.isAuthor(keys.id)`, otherwise an error is thrown. That's because we now have `ssbKeys.generate(null, null, 'bendybutt-v1')`, we can assume that the resulting `keys.id` is correctly a bendybutt SSB URI.

Reply:
done